### ISSUE TITLE
The Theano backend should not clip output values by default for the binary_crossentropy function.

### ISSUE THREAD

**Issue by airalcorn2** [fchollet/keras#4744]
The [`binary_crossentropy`](https://github.com/fchollet/keras/blob/5e75b8506c25f384311722f35b5dbbc3dd85f3fe/keras/backend/theano_backend.py#L1156) function in the Theano backend clips the output values before passing them to Theano's `binary_crossentropy` function, which results in negative model predictions (which shouldn't be occurring) producing a positive number for the cost instead of a `nan`. In theory, the model should not be generating negative predictions if the `binary_crossentropy` loss function is being used appropriately, but there is currently no check in place (on the model side) to ensure the output lies between zero and one. I learned this the hard way by forgetting to add a sigmoid activation function to my final `Dense` layer and then training a model for several days without error, only to then discover that the model generates negative predictions. The Keras implementation of `binary_crossentropy` behaves differently from the Theano `binary_crossentropy` function, which does not clip the predictions, so even though I was checking for `nan`s while training, the model never produced them. I would prefer Keras either not enable output clipping by default, or implement a check such that if the model is using a `binary_crossentropy` loss function, it will make sure the output is being passed through an activation function that produces predictions that should lie between zero and one.

```
import numpy as np
import theano.tensor as T

_EPSILON = 10e-8

preds = np.array([-0.012, -0.002, -0.007])
target = np.array([0, 0, 1])
output = preds
output = T.clip(output, _EPSILON, 1.0 - _EPSILON)
clipped = sum(T.nnet.binary_crossentropy(output, target).eval()) # The Keras loss.
not_clipped = sum(T.nnet.binary_crossentropy(preds, target).eval()) # The Theano loss.
```

**Comment by airalcorn2** [comment]
I believe the `from_logits` parameter is meant to be a type of "check", but, as far as I can tell, it doesn't look like it's ever used when `binary_crossentropy` is called. Indeed, when running the following code, negative predictions are produced by the model, but the Keras loss is equal to the `clipped` loss, not the sigmoid loss.

```
import numpy as np
import theano.tensor as T

np.random.seed(0)

from keras import backend as K
from keras.layers import Dense
from keras.models import Sequential

_EPSILON = 10e-8

model = Sequential()
model.add(Dense(1, input_dim = 3))
model.compile(optimizer = "adadelta", loss = "binary_crossentropy", metrics = ["accuracy"])

x = np.array([[-1, -2, -3]])
y = np.array([1])

preds = model.predict([x])
target = y

# Clipped loss.
output_clipped = T.clip(preds, _EPSILON, 1.0 - _EPSILON)
clipped_loss = sum(T.nnet.binary_crossentropy(output_clipped, target).eval())[0]

# Sigmoid loss.
output_sig = T.nnet.sigmoid(preds)
sigged_loss = sum(T.nnet.binary_crossentropy(output_sig, target).eval())[0]

# Theano loss.
theano_loss = sum(T.nnet.binary_crossentropy(preds, target).eval())[0]

# Keras loss.
history = model.fit([x], y, nb_epoch = 1, verbose = 0)
keras_loss = history.history["loss"][0]

print("clipped: {0:.4f}\tsigmoid: {1:.4f}\ttheano: {2:.4f}\tkeras: {3:.4f}".format(clipped_loss, sigged_loss, theano_loss, keras_loss))
# clipped: 16.1181	sigmoid: 2.0646	theano: nan	keras: 16.1181
```

**Comment by fchollet** [comment]
Worth noting that the primary function of the clipping is to preserve numeric stability. Without this clipping your models are highly likely to NaN even if they are only doing a valid series of operation.

**Comment by dontloo** [comment]
@fchollet The clipped version prevents the gradient of small values, which sometimes will also cause numerical problems when classifying a great amount of categories, and it's hard to find where the problem is without looking at the source code, so I guess it is not unnecessary to provide a clip-free version. :)

**Comment by airalcorn2** [comment]
Yes, I agree with @dontloo. I understand the purpose of clipping, but I think the decision to clip values (and at what extremes) should be left to the user. I've implemented a Clip layer in PR #4778 to hopefully alleviate this issue.

**Comment by nouiz** [comment]
I think Theano back-end behavior to propagate the nan is the right
behavior. Normally nan are propagating, doing the oposite is
conter-intuitive/conter-usual interface.

Is there a reason tensorflow does differently?

**Comment by redst4r** [comment]
must have been a strange bug in tensorflow `0.11.0rc2` (tf.clip_by_value() or the min/max functions called there); in 0.12 it's fixed and propagates the Nans as we would expect it!
